docs/design_docs/function_abi.md

```diff
@@ -1,51 +1,157 @@
+
+#### Structure
+
+A `Structure` is a common enough entity to have a dedicated calling convention. In C-like languages, this may just be a `struct`. In Python, it is typically a `dict` with an associated schema providing a name and type bound for each of its slots. In both, its slots are of fixed arity.
+
+In this convention, such a structure is represented as a `Tuple` in the native calling convention (i.e. `!iree.list` of variant type). The order of the elements of the tuple are the natural order of the structure, where that is either:
+
+- For a C-like system where order is determinate, it is the order of declaration.
+- For a name-based system (i.e. bind to `dict`) where no order is defined, the natural order will be the lexically sorted order of the keys.
+
+#### Encoded String
+
+Most languages interop between byte arrays (i.e. the native ABI `String` type) by way of applying an encoding. Such strings are just represented as a `ByteString` in the native ABI (i.e. `!iree.list<i8>`).
+
+#### Typed List
+
+High level lists which all share the same type bound are represented as a `TypedList` in the native ABI.
+
+#### NDArray of Reference Types
+
+NDArrays of reference types are considered separately from those of value types. Internally, the code generated for them is completely different from what gets generated for numeric based arrays (i.e. has ref-counting, ownership semantics, non-POD, etc). These types are permitted for completeness, not necessarily performance: by nature they are already indirected and have overheads.
+
+In the native ABI, these are represented as a composite tuple type (i.e. today a list since sugar for tuple is not yet defined): `!iree.tuple<!iree.list<T>, !iree.list<index>>`. The first element of the tuple is the list of values, packed with a C-Layout and the second element is the list of dimension sizes.
+
+#### Reflection
+
+Additional reflection metadata may be encoded in a custom JSON form, providing additional typing hints for arguments and results. If present, this will be a reflection attribute with key `d`, containing a serialized JSON object.
+
+The JSON object contains:
+
+- `a` (array): List of type records for each argument.
+- `r` (array): List of type records for each argument.
+
+Type records are one of:
+
+- A string naming a primitive type:
+
+  - `i[0-9]+`: Integer type with given bit width
+  - `f[0-9]+`: IEEE floating point type with given bit width
+  - `bf16`: BFloat16
+
+- `null`: A null reference value
+- `unknown`: An unknown/unmapped type
+- An array, interpreted as a tuple describing a compound type.
+
+##### Compound type tuples
+
+A compound type tuple has a type identifier as its first element, followed with type specific fields:
+
+- `["ndarray", {element_type}, {rank}, {dim...}]`: For unknown rank, the `rank` will be `null` and there will be no dims. Any unknown dim will be `null`.
```

Comment:
I think that is what I intended (and am doing in the code). Did I spell it wrong?

Reply:
ahh I interpreted the usage as above in string form (`"null"` `"undefined"`) mostly because JSON is JSON and `undefined` is not a valid token (there's no way to differentiate between null and undefined in JSON). This won't be a problem internally for the compiler if we use MLIR attrs and I'm not too concerned about any of it if the JSON stuff is a semi-stable dynamic-language-bindings-specific thing.